fn/context_guard.go

```diff
@@ -11,103 +12,286 @@ var (
+// the timeout is reached or, if the Blocking option is not provided, the main
+// quit signal is triggered.
+func TimeoutCGOpt() ContextGuardOption {
+	return func(o *ctxGuardOptions) {
+		o.withTimeout = true
+		o.timeout = DefaultTimeout
+	}
+}
+
+// WithCtx is used to derive a cancellable context from the parent. Various
+// options can be provided to configure the behaviour of the derived context.
+func (g *ContextGuard) WithCtx(ctx context.Context,
+	options ...ContextGuardOption) (context.Context, context.CancelFunc) {
+
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	// Exit early if the parent context has already been cancelled.
+	select {
+	case <-ctx.Done():
+		return ctx, func() {}
+	default:
+	}
+
+	var opts ctxGuardOptions
+	for _, o := range options {
+		o(&opts)
+	}
+
+	if opts.blocking {
+		if opts.withTimeout {
+			return g.withCtxBlockingTimeout(ctx, opts.timeout)
 		}
-	}()
+
+		return g.withCtxBlocking(ctx)
+	}
+
+	// If the call is non-blocking, then we can exit early if the main quit
+	// signal has been triggered.
+	select {
+	case <-g.quit:
+		ctx, cancel := context.WithCancel(ctx)
+		cancel()
+
+		return ctx, cancel
+	default:
+	}
+
+	if opts.withTimeout {
+		return g.withCtxQuitTimeout(ctx, opts.timeout)
+	}
+
+	return g.withCtxQuit(ctx)
+}
+
+// withCtxBlockingTimeout is used to derive a cancellable context with a timeout
+// that will NOT be cancelled if the main quit signal is triggered, to block
+// shutdown of important tasks. The context will be cancelled if the timeout is
+// reached.
+func (g *ContextGuard) withCtxBlockingTimeout(ctx context.Context,
+	timeout time.Duration) (context.Context, func()) {
+
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+
+	g.ctxBlocking(ctx, cancel)
 
 	return ctx, cancel
 }
 
-// CtxBlocking is used to create a cancellable context that will NOT be
-// cancelled if the main quit signal is triggered, to block shutdown of
-// important tasks. The context will be cancelled if the timeout is reached.
-func (g *ContextGuard) CtxBlocking() (context.Context, func()) {
-	return g.CtxBlockingCustomTimeout(g.DefaultTimeout)
+func (g *ContextGuard) withCtxBlocking(ctx context.Context) (context.Context,
+	func()) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	g.ctxBlocking(ctx, cancel)
+
+	return ctx, cancel
 }
 
-// CtxBlockingCustomTimeout is used to create a cancellable context with a
-// custom timeout that will NOT be cancelled if the main quit signal is
-// triggered, to block shutdown of important tasks. The context will be
-// cancelled if the timeout is reached.
-func (g *ContextGuard) CtxBlockingCustomTimeout(
-	timeout time.Duration) (context.Context, func()) {
+// withCtxQuitTimeout is used to derive a cancellable context that will be cancelled
+// if the main quit signal is triggered, the parent context is cancelled or the
+// timeout is reached.
+//
+// NOTE: the caller must hold the ContextGuard's mutex before calling this
```

Comment:
nit: since this is now only used in a single place (and is pretty small),  we could inline this in the main method above. Then holding the lock is guaranteed. Same with `withCtxQuit()`.

Reply:
ah indeed! that simplifies things quite a bit! 🙏